File: cpp/030.cpp
/*
The important part of the code is determining the upper limit of the brute force, taking the digit 9 into consideration, the highest number will be X*9^5 where X is the number of possible digits, also as we note that 9^5 = 59049, we see 5 minumum digits for maximum sum, but then 5*9^5 gives us 6 digits, i.e. 6*9^5 is our upper limit, beyond which the number of digits shall not match sums of power of 9 meaning the number will never be the same after such a point, therefore the value 6*9^5 = 354294 will be our upper limit.
 */

#include <iostream>
#include <cmath>

using namespace std;

bool fifth(long int num)    //Function flags all values that are equal to the sum of their digits to power of fifth.
{
    long int dig, orig, sum = 0;
    orig = num;
    while (num > 0)
    {
        dig = num % 10;
        sum = sum + pow(dig, 5);
        num = num/10;
    }
    return (sum == orig);   //Return TRUE only if original number satisfies condition.
}

int main()
{
    long int finalsum = 0;
    for (long int i=2; i < 354294; i++)     //Limit put in the loop.
    {
        if (fifth(i))
        {
            finalsum += i;  //Final sum incremented.
        }
    }
    cout << "Final sum of all such numbers is: " << finalsum;
    return 0;
}


        

File: python/030.py
'''
Problem Statement:

Surprisingly there are only three numbers that can be written as the sum of fourth powers of their digits:

1634 = 14 + 64 + 34 + 44
8208 = 84 + 24 + 04 + 84
9474 = 94 + 44 + 74 + 44
As 1 = 14 is not a sum it is not included.

The sum of these numbers is 1634 + 8208 + 9474 = 19316.

Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.
'''

n = 5
s = 0

for num in range(10, 10**(n+1)-1):
	val = sum([int(j)**n for j in str(num)])
	if val == num:
		s += val

print(s)